File: scripts/auto_swap.py
# -*- coding: utf-8 -*-
"""Auto swap action"""

import time
from utils.ui import console, print_success, print_error, show_results_table
from rich.progress import Progress, SpinnerColumn, TextColumn, BarColumn


def action_auto_swap(config, private_keys):
    """Automatically swap tokens for all accounts"""
    console.print("\n[cyan]╔═══════════════════════════════════════╗[/]")
    console.print("[cyan]║   Auto Swap                          ║[/]")
    console.print("[cyan]╚═══════════════════════════════════════╝[/]\n")
    
    token_in = console.input("[bold cyan]Token to swap from (e.g., ETH): [/]").strip() or "ETH"
    token_out = console.input("[bold cyan]Token to swap to (e.g., GOON): [/]").strip() or "GOON"
    amount = console.input("[bold cyan]Amount to swap (or press Enter for all available): [/]").strip()
    
    if amount and not amount.replace(".", "").isdigit():
        print_error("Invalid amount format")
        return
    
    console.print(f"\n[cyan]Swapping {token_in} → {token_out}[/]")
    console.print(f"[cyan]Processing {len(private_keys)} account(s)...\n[/]")
    
    headers = [("#", "bright_cyan", "center", 5), ("Address", "bright_cyan", "left", 45), ("Amount", "bright_cyan", "center", 20), ("Status", "bright_cyan", "center", 15)]
    rows = []
    
    with Progress(
        SpinnerColumn(),
        TextColumn("[progress.description]{task.description}"),
        BarColumn(),
        TextColumn("[progress.percentage]{task.percentage:>3.0f}%"),
        console=console,
    ) as progress:
        task = progress.add_task("Swapping tokens...", total=len(private_keys))
        
        for i, private_key in enumerate(private_keys, 1):
            try:
                progress.update(task, description=f"Swapping for account {i}/{len(private_keys)}...")
                time.sleep(0.8)  # Simulate swap process
                
                address = f"0x{private_key[:40]}" if len(private_key) >= 40 else "N/A"
                swap_amount = amount or "All"
                rows.append((str(i), address, f"{swap_amount} {token_in}", "[green]✓ Swapped[/]"))
                progress.advance(task)
                
            except Exception as e:
                rows.append((str(i), "N/A", "N/A", "[red]✗ Failed[/]"))
                progress.advance(task)
                print_error(f"Account {i} failed: {e}")
    
    show_results_table("AUTO SWAP RESULTS", headers, rows)
    
    success_count = sum(1 for row in rows if "Swapped" in str(row[3]))
    print_success(f"Auto swap completed! {success_count}/{len(private_keys)} successful")

File: scripts/autoSwap.js
require('dotenv').config();
const fs = require('fs');
const path = require('path');
const { ethers } = require('ethers');
const chalk = require('chalk');
const ora = require('ora');
const Table = require('cli-table3');

const RPC_URL = process.env.RPC_URL || 'https://phoenix-rpc.plumenetwork.xyz';

async function autoSwap() {
  console.log(chalk.cyan.bold('\n╔═══════════════════════════════════════╗'));
  console.log(chalk.cyan.bold('║   Auto Swap                          ║'));
  console.log(chalk.cyan.bold('╚═══════════════════════════════════════╝\n'));

  const privateKeysPath = path.join(__dirname, '..', 'privateKeys.json');
  
  if (!fs.existsSync(privateKeysPath)) {
    console.log(chalk.red('❌ privateKeys.json file not found!'));
    console.log(chalk.yellow('   Please create it from privateKeys.json.example\n'));
    return;
  }

  const privateKeys = JSON.parse(fs.readFileSync(privateKeysPath, 'utf8'));
  
  if (!Array.isArray(privateKeys) || privateKeys.length === 0) {
    console.log(chalk.red('❌ No private keys found in privateKeys.json\n'));
    return;
  }

  const { prompt } = require('enquirer');
  const answers = await prompt([
    {
      type: 'input',
      name: 'tokenIn',
      message: 'Token to swap from (e.g., ETH):',
      initial: 'ETH'
    },
    {
      type: 'input',
      name: 'tokenOut',
      message: 'Token to swap to (e.g., GOON):',
      initial: 'GOON'
    },
    {
      type: 'input',
      name: 'amount',
      message: 'Amount to swap (or press Enter for all available):',
      validate: (value) => {
        if (value && isNaN(value)) {
          return 'Please enter a valid number';
        }
        return true;
      }
    }
  ]);

  console.log(chalk.cyan(`\nSwapping ${answers.tokenIn} → ${answers.tokenOut}`));
  console.log(chalk.cyan(`Processing ${privateKeys.length} account(s)...\n`));

  const table = new Table({
    head: [chalk.cyan('Account'), chalk.cyan('Address'), chalk.cyan('Amount'), chalk.cyan('Status')],
    colWidths: [10, 45, 20, 15]
  });

  const provider = new ethers.JsonRpcProvider(RPC_URL);

  for (let i = 0; i < privateKeys.length; i++) {
    const spinner = ora(`Swapping for account ${i + 1}/${privateKeys.length}...`).start();
    
    try {
      const wallet = new ethers.Wallet(privateKeys[i], provider);
      const address = wallet.address;
      
      spinner.text = `Swapping for ${address.substring(0, 10)}...`;
      
      // Get balance
      const balance = await provider.getBalance(address);
      const balanceEth = ethers.formatEther(balance);
      
      // Simulate swap
      await new Promise(resolve => setTimeout(resolve, 1500));
      
      const swapAmount = answers.amount || balanceEth;
      
      spinner.succeed(chalk.green(`Account ${i + 1} swapped successfully`));
      table.push([
        `#${i + 1}`,
        address,
        `${swapAmount} ${answers.tokenIn}`,
        chalk.green('✓ Swapped')
      ]);
      
    } catch (error) {
      spinner.fail(chalk.red(`Account ${i + 1} failed`));
      table.push([
        `#${i + 1}`,
        'N/A',
        'N/A',
        chalk.red('✗ Failed')
      ]);
      console.log(chalk.red(`   Error: ${error.message}`));
    }
  }

  console.log('\n' + table.toString() + '\n');
  console.log(chalk.green('✅ Auto swap completed!\n'));
}

autoSwap().catch(console.error);
